xapian-letor/ranker/ranker.cc

```diff
@@ -192,32 +192,29 @@ write_to_file(const std::vector<Xapian::FeatureVector> & list_fvecs, const strin
     }
 }
 
+// Query file is in the format: <qid> '<query_string>'
 static std::pair<string, string>
 parse_query_string(const string & query_line, int line_number)
 {
```

Comment:
I had to think twice about why this will work (a) when `query_line` is an empty string, and (b) when `i` is `::npos`. I don't think we should change the code, because it's quite tight as it is, but perhaps some more comments would help future readers.

I _do_ think we should have explicit tests for an empty line, a line without a space, and perhaps a line that ends with a space but has no single quote at all. At least that would give someone glancing at this confidence that the code does cope with all the edge cases.

Reply:
The case of an empty string will never reach this part of the code as it will be eliminated on line 265 in ranker.cc. As for the remaining cases, I have added corresponding query files.

Reply:
>  perhaps a line that ends with a space but has no single quote at all.
>
xapian-check-patch is complaining about trailing whitespace in this case.

Reply:
If it is never supposed to cope with an empty string, then put an assert in there to state that clearly.

Reply:
Oh, and equivalent to a no-quote line ending with a space would be something like:

```
12345 noquote
```

I believe that will test the same code path…

Reply:
fixed in a fixup commit.